modules/forwarder/iam.tf

```diff
@@ -0,0 +1,92 @@
+resource "aws_iam_role" "this" {
+  name               = var.name
+  assume_role_policy = data.aws_iam_policy_document.lambda_assume_role_policy.json
+
+  dynamic "inline_policy" {
+    for_each = {
+      for k, v in {
+        logging = data.aws_iam_policy_document.logging.json
+        writer  = data.aws_iam_policy_document.writer.json
+        queue   = data.aws_iam_policy_document.queue.json
+        reader  = length(var.source_bucket_names) > 0 ? data.aws_iam_policy_document.reader.json : null
+      } : k => v if v != null
+    }
+
+    content {
+      name   = inline_policy.key
+      policy = inline_policy.value
+    }
+  }
+}
+
+data "aws_iam_policy_document" "lambda_assume_role_policy" {
+  statement {
+    effect  = "Allow"
+    actions = ["sts:AssumeRole"]
+    principals {
+      type        = "Service"
+      identifiers = ["lambda.amazonaws.com"]
+    }
+  }
+}
+
+data "aws_iam_policy_document" "logging" {
+  statement {
+    effect = "Allow"
+    actions = [
+      "logs:CreateLogGroup",
+      "logs:CreateLogStream",
+      "logs:PutLogEvents",
+    ]
+
+    resources = [
+      "${aws_cloudwatch_log_group.this.arn}*",
+    ]
+  }
+}
+
+data "aws_iam_policy_document" "writer" {
+  statement {
+    actions = [
+      "s3:PutObject",
+      "s3:PutObjectTagging",
+    ]
+
+    resources = ["*"]
```

Comment:
is this to write the files to the s3 bucket for filedrop? should this need access to all buckets or just the filedrop bucket?

Reply:
It ends up not having wildcard access because we lock it down in the subsequent condition. We can't really lock down by "bucket" when dealing with an access point, because all we have is a bucket alias. IAM is one of the places where aliases cannot be used interchangeably with bucket names https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-points-alias.html